### PR TITLE
accounts/abi/bind: support event filtering in abigen

### PR DESCRIPTION
**Pull request by karalabe**
Quite a long time ago we landed support for native [Go bindings for Ethereum contracts](https://github.com/ethereum/go-ethereum/wiki/Native-DApps:-Go-bindings-to-Ethereum-contracts). Back then we figured events will come very soon, but they turned out to require a complete overhaul of our subscription system, filtering system, etc. In short, it never happened.

This PR is fulfilling that promise made long ago, expanding `abigen` with support for strongly typed, native event filtering for Ethereum contracts from within Go!

Supersedes https://github.com/ethereum/go-ethereum/pull/15246 which only started work towards event filtering, but did not generate typed Go code. Supersedes https://github.com/ethereum/go-ethereum/pull/15832 which didn't support filtering at all, just full retrieval.

---

This change is both-way incompatible:

 * Bindings generated prior to this version will need to be regenerated due to modifications made to the base layer on top of which `abigen` contracts are running.
 * Bindings generated with the new `abigen` will require the underlying supporting code (`go-ethereum`) to be updated too as the new code will depend on newly added features.

**Existing client code will not break as no client API changes have been made, only extensions!**

---

As a recap, `abigen` could take a Solidity source file (or ABI spec with an optional EVM bytecode) and generate a Go wrapper out of it. For out original code [token.sol](https://gist.github.com/karalabe/08f4b780e01c8452d989), this looked like [token.go](https://gist.github.com/karalabe/c22aab73194ba7da834ab5b379621031).

The highlights were automatically generated types and methods to fully interact with the `Token` contract described in the Solidity file (sample API snippet):

```go
// DeployMyToken deploys a new Ethereum contract, binding an instance of MyToken to it.
DeployMyToken(auth *bind.TransactOpts, backend bind.ContractBackend, initialSupply *big.Int, tokenName string, decimalUnits uint8, tokenSymbol string) (common.Address, *types.Transaction, *MyToken, error)

// BalanceOf is a free data retrieval call binding the contract method 0x70a08231.
//
// Solidity: function balanceOf( address) constant returns(uint256)
func (_MyToken *MyToken) BalanceOf(opts *bind.CallOpts, arg0 common.Address) (*big.Int, error) {

// Transfer is a paid mutator transaction binding the contract method 0xa9059cbb.
//
// Solidity: function transfer(_to address, _value uint256) returns()
func (_MyToken *MyToken) Transfer(_to common.Address, _value *big.Int) (*types.Transaction, error) {
```

These generated methods could be used to deploy new contracts, bind existing ones, call methods on the contracts to read its state, and initiate state modifying transactions, all from within Go with strongly typed method arguments and return values.

---

This PR expands the generated code with new types and methods to support filtering past contract events, and subscribing to a stream of future events!

For every event present in the contract, `abigen` will create a Go counterpart. E.g. for the Transfer event in the token contract:

```sol
event Transfer(address indexed from, address indexed to, uint256 value);
```

```go
// MyTokenTransfer represents a Transfer event raised by the MyToken contract.
type MyTokenTransfer struct {
	From  common.Address
	To    common.Address
	Value *big.Int
	Raw   types.Log // Blockchain specific contextual infos
}
```

For each event, `abigen` will generate a `Filter<event name>` method to retrieve past logs. As you can see below, the topics being filtered for are strongly typed Go types, **not topic hashes as all other APIs surface**. This allows calling code to be meaningfully read and understood without having to guess what some cryptic numbers mean. `abigen` will generate all the necessary code to convert the user types into topic filter criteria under the hood.

The log filterer method returns an iterator that can be used to iterate over the found logs, each already unpacked from its raw Ethereum RLP encoded form (and topic form) into strongly typed Go structs like `MyTokenTransfer` listed above.

```go
// FilterTransfer is a free log retrieval operation binding the contract event 0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef.
//
// Solidity: event Transfer(from indexed address, to indexed address, value uint256)
func (_MyToken *MyToken) FilterTransfer(opts *bind.FilterOpts, from []common.Address, to []common.Address) (*MyTokenTransferIterator, error) {
```

Similarly, for each event, `abigen` will generate a `Watch<event name>` method to subscribe to future logs and similarly to filtering past events, subscribing to future ones can be done via strongly typed criteria.

The method also takes a sink channel as an argument to deliver newly found contract events on, and returns a subscription which can be used to tear down the watcher constructs. As with past event subscriptions, the events streamed in the user-provided channel are strongly typed like `MyTokenTransfer` above.

```go
// WatchTransfer is a free log subscription operation binding the contract event 0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef.
//
// Solidity: event Transfer(from indexed address, to indexed address, value uint256)
func (_MyToken *MyTokenFilterer) WatchTransfer(opts *bind.WatchOpts, sink chan<- *MyTokenTransfer, from []common.Address, to []common.Address) (event.Subscription, error) {
```

The full generated code with this PR's `abigen` is available at [token.go](https://gist.github.com/karalabe/1fc6fec41a7e59a043eb8f614afce4fa).

# Caveats

Filtering for past log events returns an iterator for future API stability. The current implementation under the hood runs the entire user filtering in one big go, caches the entire results and then uses the iterator to simulate streaming filtering. This will be replaced eventually, but to avoid breaking the API then, we're enforcing this future mode of operation in user code already now.

Watching for future log events has an optional parameter for specifying the starting block number. This is currently a noop as `go-ethereum` does not yet support subscribing to past events, but when support lands for it, `abigen` will naively be able to use it without API changes.

### CODE REVIEW

**Comment by obscuren** [comment]
Long time coming.  Well done Peter 👏🏼

File: accounts/abi/bind/backend.go [hunk 1]
```diff
@@ -90,8 +84,29 @@ type ContractTransactor interface {
 	SendTransaction(ctx context.Context, tx *types.Transaction) error
 }
 
+// ContractFilterer defines the methods needed to accesslog events using a one-off
```
Comment (ligi) [inline accounts/abi/bind/backend.go]: -a

File: accounts/abi/bind/backends/simulated.go [hunk 2]
```diff
@@ -302,7 +314,69 @@ func (b *SimulatedBackend) SendTransaction(ctx context.Context, tx *types.Transa
...
+// subscription immediately, which can be used to stream the found events.
+func (b *SimulatedBackend) SubscribeFilterLogs(ctx context.Context, query ethereum.FilterQuery, ch chan<- types.Log) (ethereum.Subscription, error) {
+	// Subscribe to contract events
+	sink := make(chan []*types.Log)
+
+	sub, err := b.events.SubscribeLogs(query, sink)
+	if err != nil {
+		return nil, err
+	}
+	// Since we're getting logs in batches, we need to flattem them into a plain stream
```
Comment (ligi) [inline accounts/abi/bind/backends/simulated.go]: flattem -> flatten

File: accounts/abi/bind/base.go [hunk 3]
```diff
@@ -225,6 +244,104 @@ func (c *BoundContract) transact(opts *TransactOpts, contract *common.Address, i
 	return signedTx, nil
 }
 
+// FilterLogs filters contract logs for past blocks, returning the necessary
+// channels to construct a strongly types bound iterator on top of them.
```
Comment (ligi) [inline accounts/abi/bind/base.go]: "strongly types bound iterator" -> "strongly typed bound iterator"

File: accounts/abi/abi.go [hunk 4]
```diff
@@ -97,7 +97,6 @@ func (abi *ABI) UnmarshalJSON(data []byte) error {
 		Type      string
 		Name      string
 		Constant  bool
-		Indexed   bool
```
Comment (gballet) [inline accounts/abi/abi.go]: I don't understand why you remove this from the list of fields to be un-marshaled? I understand  that it is derived from another source, but what if the field is still present in some client code?

Comment (karalabe) [reply accounts/abi/abi.go]: Indexed only makes sense in the context of an event field (which is inside "inputs"). In the outer ABI context there's no notion of "indexed" afaik.

File: accounts/abi/bind/base.go [hunk 5]
```diff
@@ -225,6 +244,104 @@ func (c *BoundContract) transact(opts *TransactOpts, contract *common.Address, i
...
+	}), nil
+
+	if err != nil {
+		return nil, nil, err
+	}
+	return logs, sub, nil
+}
+
+// WatchLogs filters subscribes to contract logs for future blocks, returning a
+// subscription object that can be used to tear down the watcher.
```
[the remark this answers is not in the document]
Comment (karalabe) [reply accounts/abi/bind/base.go]: E.g. When I'm filtering for past events, my code will usually block waiting for the next past event to arrive, until I consumed all past events I filtered for and the subsystem tells me there are no more events coming.

When I'm filtering for future events, my code will not block, rather be event driven, waiting on multiple possible event sources for data to arrive and react to.

---

To be honest, I can't think of an application where I would want to handle past and future events the same way. E.g. I can filter for past Akasha events to retrieve the stuff the user posted and serve on a restful API, but that needs to stop and return, not block for arbitrary future events. On the other hand, I can filter for future Akasha events to detect when users post new content and cache them in my API server via IPFS immediately during posting while the user is still online. The two use cases are wildly different and have different requirements.

[the remark this answers is not in the document]
Comment (gballet) [reply accounts/abi/bind/base.go]: > To be honest, I can't think of an application where I would want to handle past and future events the same way.

I was thinking that implementing something like `tail -f` in Unix would be very complicated with this approach. I guess that this is an unlikely use case, then.

[the remark this answers is not in the document]
Comment (karalabe) [reply accounts/abi/bind/base.go]: The latter `WatchXYZ` is the one which supposed to handle that case.

**Comment by shoenseiwaso** [comment]
Great work, thank you!

Presumably if I want to iterate through *all* events in the order they are received (as opposed to *per-event-type* ordering), I would still need to use the `FilterLogs()`/`SubscribeFilterLogs()` => `select { ... }` => `UnpackLog()` (<== new and supports indexed events now, thank you!) ?

**Comment by dt665m** [comment]
I know this is a pretty new PR but is there any documentation or usage examples of watching events?  Haven't been able to get it to work.  Does this support 'geth --light'?

**Comment by writetosalman** [comment]
A code example will really help here. Thanks.

**Comment by ArtMartiros** [comment]
# Working example 
## My event
event YearChanged(uint32 indexed year);

## Past events
	opt := &bind.FilterOpts{}
	s := []uint32{}
	past, err := contract.FilterYearChanged(opt, s)
	if err != nil {
		log.Fatalf("Failed FilterYearChanged: %v", err)
	}
	notEmpty := true
	for notEmpty {
		notEmpty = past.Next()
		if notEmpty {
			fmt.Println("event log:", past.Event.Year)
		}
	}

## Listen future events 

	var blockNumber uint64 = 2900000
	s := []uint32{}
	ch := make(chan *contract.EventTesterYearChanged)
	opts := &bind.WatchOpts{}
	opts.Start = &blockNumber
	_, err := contract.WatchYearChanged(opts, ch, s)
	if err != nil {
		log.Fatalf("Failed WatchYearChanged: %v", err)
	}
	var newEvent *contract.EventTesterYearChanged = <-ch
	fmt.Println(newEvent.Year)

**Comment by hossein761** [comment]
I am trying to to fetch some past events from a contract as follows. In my example I have used abigen to create bindings for ZIL:

`conn, err := ethclient.Dial("https://mainnet.infura.io/ACCESS_TOKEN")
	if err != nil {
		log.Fatal("Error connecting to the Ethereum client: ", err)
	}

	zil, err := bindings.NewZIL(common.HexToAddress("0x05f4a42e251f2d52b8ed15e9fedaacfcef1fad27"), conn)
	if err != nil {
		log.Fatal("Failed to instantiate a ZIL contract")
	}
	opz :=  &bind.CallOpts{}
	s, err := zil.Name(opz)
	if err != nil {
		log.Error("some error", err)
	}

	log.Info("name: ", s) // here the name of the contract is correctly printed


       // I want to filter between start and end blocks
	end := uint64(4388365)
	opts := &bind.FilterOpts{
		Start: 	4288360,
		End: &end,
	}
	transferEvts, err := zil.FilterTransfer(opts, []common.Address{}, []common.Address{})
	if err != nil {
		log.Fatal("Error creating filter on Transfer events", err.Error())
	}

	notEmpty := true
	for notEmpty {
			notEmpty = transferEvts.Next()
			if notEmpty {
				log.Info("Transfer: ", transferEvts.Event.Value)
			} else {
				log.Info("No more events ", transferEvts.Error())
			}
	}` 

This follows the example above given by @ArtMartiros . The name of the contract is printed so I am sure the connection works. However, no transaction is printed and transferEvts.Error() is empty.
Can someone please help me with this?

**Comment by ArtMartiros** [comment]
@hossein761  your code looks correct.

**Comment by dimiandre** [comment]
How can we handle connection lost ?

actually i have a code like this:

`
	// Setup Watch options (none)
	opts := &bind.WatchOpts{}

	// Create a new streaming channel
	ch := make(chan *contract.TokenTransfer)
	quit := make(chan struct{})

	// attach the channel to a new watcher
	sub, err := contract.WatchTransfer(opts, ch, []common.Address{}, []common.Address{})
	if err != nil {
		log.Fatalf("Failed Watch Transfers: %v", err)
	}

	for sub != nil {
		select {
		case event := <-ch:
			go sendToPortal(event.Raw, event.From.Hex(), event.To.Hex(), *event.Value)
		
		case <-quit:
			fmt.Println("Subscription break")
			return
		}
	}`

maybe i'm doing something wrong, but quit is never fired. Even if i loose connection to the node. 

also would be cool to use the event.Resubscribe function as parameter directly in abigen, what do you think ?

**Pull request merged**